### PR TITLE
fix(lsp): allow buf_request_all handler to handle missing client_id

### LINKED ISSUE #15174: `lsp.buf_request_all` can't handle unsupported method error
- `nvim --version`: v0.6.0-dev+64-g754ac5a83
- language server name/version: tsserver 4.3.4 (but same issue occurs with any language server)
- Operating system/version: macOS 11.4

<details>
<summary>nvim -c ":checkhealth nvim lspconfig"</summary>

health#nvim#check
========================================================================
## Configuration
  - OK: no issues found

## Performance
  - INFO: Build type: Debug
  - WARNING: Non-optimized (DEBUG) build. Nvim will be slower.
    - ADVICE:
      - Install a different Nvim package, or rebuild with `CMAKE_BUILD_TYPE=RelWithDebInfo`.
      - https://github.com/neovim/neovim/wiki/FAQ

## Remote Plugins
  - OK: Up to date

## terminal
  - INFO: key_backspace (kbs) terminfo entry: key_backspace=\177
  - INFO: key_dc (kdch1) terminfo entry: key_dc=\E[3~
  - INFO: $TERM_PROGRAM='tmux'
  - INFO: $COLORTERM='truecolor'

## tmux
  - OK: escape-time: 0
  - INFO: Checking stuff
  - OK: focus-events: on
  - INFO: $TERM: tmux-256color

health#lspconfig#check
========================================================================
## Checking language server protocol configuration
  - INFO: null-ls: configuration checked.
  - INFO: tsserver: configuration checked.
  - INFO: sumneko_lua: configuration checked.
</details>

<details>
<summary>lsp.log</summary>
[ WARN ] 2021-07-24T16:57:45+0900 ] /usr/local/share/nvim/runtime/lua/vim/lsp.lua:87 ]	"method textDocument/codeAction is not supported by any of the servers registered for the current buffer"
</details>

### Steps to reproduce using nvim -u minimal_init.lua
The easiest way to reproduce the issue is to modify `minimal_init.lua` to start `tsserver` (or any language server) with the following config:
```lua
nvim_lsp.tsserver.setup {
    on_attach = function(client) 
        client.resolved_capabilities.code_action = false
    end,
    -- just to make sure the server attaches in in a random test directory
    root_dir = function() return vim.fn.getcwd() end
}
```
Steps: 
```
nvim -u minimal_init.lua test.ts
:lua vim.lsp.buf.code.action()
```

### Actual behaviour
Neovim shows the following error message:
```
E5108: Error executing lua /usr/local/share/nvim/runtime/lua/vim/lsp.lua:1327: table index is nil
```

### Expected behaviour
Neovim shows an error message indicating that no server can handle the given request.

This issue occurs not just with `tsserver` but in any situation where a) a language server is attached to a buffer, and b) that language server cannot handle a given request sent through `lsp.buf_request_all`, either because of its declared capabilities or because the capability was specifically disabled, as here.

I believe that the issue is caused by the fact that in this situation, `lsp.buf_request` [calls `handler` with only 3 arguments](https://github.com/neovim/neovim/blob/754ac5a834cb82f73ca552a9a55b90a40f9367f6/runtime/lua/vim/lsp.lua#L1286), but `lsp.buf_request_all` [expects the 4th argument to contain the client's ID](https://github.com/neovim/neovim/blob/754ac5a834cb82f73ca552a9a55b90a40f9367f6/runtime/lua/vim/lsp.lua#L1326) (and doesn't check whether it is `nil` before attempting to use it to index `request_results`).

I'm not sure what would be the best solution but am happy to put in a PR to fix it if I can get pointed in the right direction. Thanks!

### PR DESCRIPTION
**Pull request by jose-elias-alvarez**
Closes #15174. 

Not 100% sure that this is the right approach, but I figured a potential solution was better than nothing. The idea is to use `-1` as an index to make sure the error isn't lost completely, but another option is to do nothing if `client_id` is `nil` and call `callback` with an empty table. 

Small note: setting `result` conditionally was necessary because `result` will correspond to `bufnr` if [`buf_request` returns an unsupported error](https://github.com/neovim/neovim/blob/192adfe99f33778a85e11fbfdceb37f347a3d235/runtime/lua/vim/lsp.lua#L1286), which causes problems for implementations that expect `nil` or a table, e.g. [`lsp.buf.code_action`](https://github.com/neovim/neovim/blob/192adfe99f33778a85e11fbfdceb37f347a3d235/runtime/lua/vim/lsp/buf.lua#L434).

### CODE REVIEW

**Comment by mfussenegger** [comment]
What do you think of adjusting `buf_request` to track which clients don't support the method, and then calling the handler for *each* of the handlers that don't support it with the right client id?

**Comment by mjlbach** [comment]
We should not be sending requests to servers that don't support a given method, not handling errors returned for the server.

**Comment by fsouza** [comment]
> What do you think of adjusting `buf_request` to track which clients don't support the method, and then calling the handler for _each_ of the handlers that don't support it with the right client id?

Wouldn't that be kind of a regression from the current behavior? IIRC, the callback is only called if _none_ of the servers support the requested method. Did that change or do I not remember correctly? 🙈 

The original idea was that skipping servers that don't support the method is natural (if you're running multiple servers, it's likely that they don't have all features and that they're being used to complement each other?), but if the user calls a method which is not supported by any method, we want to report that to the callback.

**Comment by mfussenegger** [comment]
To clarify: My suggestion was to keep sending the request only to the server who support it, and not to any of the others.

The handlers would also only be called if no client supported the method.
Something like this:


```
diff --git a/runtime/lua/vim/lsp.lua b/runtime/lua/vim/lsp.lua
index f5aefc826..6fed5b6e4 100644
--- a/runtime/lua/vim/lsp.lua
+++ b/runtime/lua/vim/lsp.lua
@@ -1260,10 +1260,11 @@ function lsp.buf_request(bufnr, method, params, handler)
     bufnr    = { bufnr, 'n', true };
     method   = { method, 's' };
     handler  = { handler, 'f', true };
   }
   local client_request_ids = {}
+  local unsupported_clients = {}

   local method_supported = false
   for_each_buffer_client(bufnr, function(client, client_id, resolved_bufnr)
     if client.supports_method(method) then
       method_supported = true
@@ -1272,20 +1273,24 @@ function lsp.buf_request(bufnr, method, params, handler)
       -- This could only fail if the client shut down in the time since we looked
       -- it up and we did the request, which should be rare.
       if request_success then
         client_request_ids[client_id] = request_id
       end
+    else
+      table.insert(unsupported_clients, client_id)
     end
   end)

   -- if has client but no clients support the given method, call the callback with the proper
   -- error message.
   if not tbl_isempty(all_buffer_active_clients[resolve_bufnr(bufnr)] or {}) and not method_supported then
     local unsupported_err = lsp._unsupported_method(method)
     handler = handler or lsp.handlers[method]
     if handler then
-      handler(unsupported_err, method, bufnr)
+      for _, client_id in pairs(unsupported_clients) do
+        handler(unsupported_err, method, bufnr, client_id)
+      end
     end
     return
   end

   local function _cancel_all_requests()
```

**Comment by mjlbach** [comment]
@mfussenegger's suggestions is the fix I would prefer :)

**Comment by jose-elias-alvarez** [comment]
Thank you for the feedback! I agree that changing `buf_request` to track unsupported clients makes sense. 

We still run into an issue with `buf_request_all`, which is that `sync_handler` expects to be called with the following argument structure:

```lua
local function _sync_handler(err, _, result, client_id)
```

But `buf_request` will call it with the following arguments, meaning that `result` will be assigned to `bufnr`:

```lua
handler(unsupported_err, method, bufnr, client_id)
```

This causes issues for handlers that depend on checking if `result` is falsy, such as [`buf.code_action`](https://github.com/neovim/neovim/blob/c818d8df349fff514eef8a529afe63e8102ca281/runtime/lua/vim/lsp/buf.lua#L434) and [Telescope's code action builtin](https://github.com/nvim-telescope/telescope.nvim/blob/69adcbca876bbd12cb2d120e403bf4377c34a5a2/lua/telescope/builtin/lsp.lua#L162) (though Telescope isn't affected by this particular issue, since it does its own check to see if a given request can be handled). 

Strictly speaking, I think it's logical that handlers should check the `error` field of each table element, but it seems like that's not the case currently. Does it make sense to set `result` to `nil` if the handler is called with an error to avoid this issue?

**Comment by mfussenegger** [comment]
> Does it make sense to set result to nil if the handler is called with an error to avoid breaking things?

Yes I think so. We should change it to follow the arguments described in `:help lsp-handler`:

`function(err, method, result, client_id, bufnr, config)`

**Comment by jose-elias-alvarez** [comment]
Great, thank you! I've restructured how the handler is called to conform to the structure described in `:help lsp-handler`. 

Also, to make sure `sync_handler` in `buf_request_all` continues to do what it's supposed to do, I've changed it so that `buf_request` returns the list of unsupported client IDs and schedules calling the handlers (meaning `buf_request_all` doesn't have to be changed at all). Let me know if that makes sense – the alternative would be to skip counting results entirely if `client_request_ids` is `nil`.

File: runtime/lua/vim/lsp.lua [hunk 1]
```diff
@@ -1274,18 +1275,24 @@ function lsp.buf_request(bufnr, method, params, handler)
...
     local unsupported_err = lsp._unsupported_method(method)
     handler = handler or lsp.handlers[method]
     if handler then
-      handler(unsupported_err, method, bufnr)
+      vim.schedule(function()
+        for _, client_id in ipairs(unsupported_clients) do
+          handler(unsupported_err, method, nil, client_id, bufnr)
+        end
+      end)
+      return unsupported_clients
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp.lua:1294]: Currently the return value is documented as:

```
--@returns 2-tuple:
---  - Map of client-id:request-id pairs for all successful requests.
---  - Function which can be used to cancel all the requests. You could instead
---    iterate all clients and call their `cancel_request()` methods.
```

I think to make the error result compatible and non-surprising we could change it to a 3 result tuple

```suggestion
      return {}, function() end, unsupported_clients
```

(This then also needs a docs update)

Comment (jose-elias-alvarez) [reply runtime/lua/vim/lsp.lua]: Makes sense!

**Comment by jose-elias-alvarez** [comment]
Updated the return values and docs, and the fix takes care of the issue. Let me know if everything looks OK.

**Comment by mfussenegger** [comment]
Now it would look good to me, but it seems as if tests are failing.

If you run:


```
make functionaltest TEST_FILE=test/functional/plugin/lsp_spec.lua 
```

You can see the failure:

```
[  ERROR   ] 1 error, listed below:
[  ERROR   ] test/functional/plugin/lsp_spec.lua @ 405: LSP basic_init test should call unsupported_method when trying to call an unsupported method
test/functional/helpers.lua:246: test/functional/plugin/lsp_spec.lua:430: attempt to index local 'lsp_cb_call' (a userdata value)

stack traceback:
	test/functional/helpers.lua:246: in function 'run_session'
	test/functional/helpers.lua:251: in function 'run'
	test/functional/plugin/lsp_spec.lua:123: in function 'test_rpc_server'
	test/functional/plugin/lsp_spec.lua:409: in function <test/functional/plugin/lsp_spec.lua:405>

```

**Comment by jose-elias-alvarez** [comment]
Right, I imagine this is because calling handlers is now scheduled. I was able to fix the test (and make sure that a related test didn't always pass) by adding a wait time before checking that the handler was called, but let me know if there's a more elegant solution.

**Comment by mfussenegger** [comment]
> Right, I imagine this is because calling handlers is now scheduled. I was able to fix the test (and make sure that a related test didn't pass unconditionally) by adding a wait time before checking that the handler was called, but let me know if there's a more elegant solution.

What's the reason that the schedule is now required?

**Comment by jose-elias-alvarez** [comment]
> > Right, I imagine this is because calling handlers is now scheduled. I was able to fix the test (and make sure that a related test didn't pass unconditionally) by adding a wait time before checking that the handler was called, but let me know if there's a more elegant solution.
> 
> What's the reason that the schedule is now required?

`buf_request_all` expects to receive a table of `client_request_ids` as a return value from `buf_request` before `sync_handler` is called, so calling the handler before returning (at a point `client_request_ids` is `nil`) would cause an error. 

It's possible to fix `buf_request_all` by having it check it `client_request_ids` is `nil` before attempting to iterate over it, but I wonder if this would cause issues elsewhere, since (from my understanding) in a normal scenario, `buf_request_all` should return before calling the handler.

**Comment by mfussenegger** [comment]
> > > Right, I imagine this is because calling handlers is now scheduled. I was able to fix the test (and make sure that a related test didn't pass unconditionally) by adding a wait time before checking that the handler was called, but let me know if there's a more elegant solution.
> > 
> > 
> > What's the reason that the schedule is now required?
> 
> `buf_request_all` expects to receive a table of `client_request_ids` as a return value from `buf_request` before `sync_handler` is called, so calling the handler before returning (at a point `client_request_ids` is `nil`) would cause an error.
> 
> It's possible to fix `buf_request_all` by having it check it `client_request_ids` is `nil` before attempting to iterate over it, but I wonder if this would cause issues elsewhere, since (from my understanding) in a normal scenario, `buf_request` should return before calling the handler.


I wonder if we should adjust `buf_request_all` to use `vim.lsp.buf_get_clients()` and the `client.request` methods. It would duplicate some logic in regards to the unsupported methods handling, but it would have the advantage that `buf_request_all`  can have control over the clients and know up-front how many results there will be - that would avoid this dependency on implementation detail in regards to when the handlers are triggered.

I think with the current change as it is in this PR, the `set_expected_result_count` logic would be broken, given that it doesn't account for the `unsupported_clients`. `expected_result_count` would remain at `0` and the `callback` would fire multiple times.

**Comment by jose-elias-alvarez** [comment]
> 
> I wonder if we should adjust `buf_request_all` to use `vim.lsp.buf_get_clients()` and the `client.request` methods. It would duplicate some logic in regards to the unsupported methods handling, but it would have the advantage that `buf_request_all` can have control over the clients and know up-front how many results there will be - that would avoid this dependency on implementation detail in regards to when the handlers are triggered.
> 
> I think with the current change as it is in this PR, the `set_expected_result_count` logic would be broken, given that it doesn't account for the `unsupported_clients`. `expected_result_count` would remain at `0` and the `callback` would fire multiple times.

I tried setting this up locally, and what I found is that having `buf_request_all` get clients and make requests does remove the timing issue, but keeping the return value (a callback that cancels all successful requests) requires duplicating most of the logic in `buf_request`. 

I just pushed a change that makes `set_expected_result_count` correctly account for unsupported clients. Let me know if that makes sense or if you prefer to make `buf_request_all` independent from `buf_request`.

**Comment by mfussenegger** [comment]
> Let me know if that makes sense or if you prefer to make buf_request_all independent from buf_request.

I think I would prefer it - or we'd need to come up with a different way to ensure the tests work. A `vim.wait(5)` is quite unreliable and could cause flaky test failures.

But maybe let's see if @mjlbach has some other opinion or more inputs.

**Comment by jose-elias-alvarez** [comment]
Separately, I think there's something weird going on with the tests to check if the handler is called. Now that we're tracking unsupported clients, I tried simplifying `buf_request` as follows (validation / comments removed for brevity):

```lua
function lsp.buf_request(bufnr, method, params, handler)
  local client_request_ids = {}
  local unsupported_clients = {}

  local method_supported = false
  for_each_buffer_client(bufnr, function(client, client_id, resolved_bufnr)
    if client.supports_method(method) then
      method_supported = true
      local request_success, request_id = client.request(method, params, handler, resolved_bufnr)

      if request_success then
        client_request_ids[client_id] = request_id
      end
    else
      table.insert(unsupported_clients, client_id)
    end
  end)

  if not method_supported and vim.tbl_count(unsupported_clients) > 0 then
    local unsupported_err = lsp._unsupported_method(method)
    handler = handler or lsp.handlers[method]
    if handler then
      for _, client_id in ipairs(unsupported_clients) do
        handler(unsupported_err, method, nil, client_id, bufnr)
      end
    end
  end

  local function _cancel_all_requests()
    for client_id, request_id in pairs(client_request_ids) do
      local client = active_clients[client_id]
      client.cancel_request(request_id)
    end
  end

  return client_request_ids, _cancel_all_requests, unsupported_clients
end
```

This works in an actual environment, but it seems that in the test, the `for_each_buffer_client` callback is never called because `active_clients` is empty (even though `all_buffer_active_clients[bufnr]` has a `client_id` from `lsp.buf_attach_client`). This means that calling the handler once for each unsupported client won't work, and the test asserting that the handler was called fails. 

I imagine this is because of how the test server is set up, but I don't know enough about the test setup to fix it cleanly. Sorry that this has gotten so complicated!

**Comment by jose-elias-alvarez** [comment]
> > Let me know if that makes sense or if you prefer to make buf_request_all independent from buf_request.
> 
> I think I would prefer it - or we'd need to come up with a different way to ensure the tests work. A `vim.wait(5)` is quite unreliable and could cause flaky test failures.

Here's an example for how `buf_request_all` might work independently of `buf_request`. As you can see, there's a lot of overlap, and I'm not sure if there's a good way to simplify it and maintain the same functionality:

```lua
function lsp.buf_request_all(bufnr, method, params, callback)
  local client_request_ids = {}
  local unsupported_clients = {}

  local request_results = {}
  local result_count = 0
  local expected_result_count = 0
  local function _sync_handler(err, _, result, client_id)
    request_results[client_id] = { error = err, result = result }
    result_count = result_count + 1

    if result_count >= expected_result_count then
      callback(request_results)
    end
  end

  local method_supported = false
  for_each_buffer_client(bufnr, function(client, client_id, resolved_bufnr)
    if client.supports_method(method) then
      method_supported = true
      local request_success, request_id = client.request(method, params, _sync_handler, resolved_bufnr)
      if request_success then
        expected_result_count = expected_result_count + 1
        client_request_ids[client_id] = request_id
      end
    else
      table.insert(unsupported_clients, client_id)
    end
  end)

  if not method_supported and vim.tbl_count(unsupported_clients) > 0 then
    expected_result_count = vim.tbl_count(unsupported_clients)
    local unsupported_err = lsp._unsupported_method(method)
    for _, client_id in ipairs(unsupported_clients) do
      _sync_handler(unsupported_err, method, nil, client_id)
    end
  end

  return function()
    for client_id, request_id in pairs(client_request_ids) do
      local client = active_clients[client_id]
      client.cancel_request(request_id)
    end
  end
end
```

> I wonder if we should adjust buf_request_all to use vim.lsp.buf_get_clients() and the client.request methods. It would duplicate some logic in regards to the unsupported methods handling, but it would have the advantage that buf_request_all can have control over the clients and know up-front how many results there will be - that would avoid this dependency on implementation detail in regards to when the handlers are triggered.

The issue is that (if I understand correctly) we don't know the expected result count until: 

1. We actually call `client.request` and know if the request was successful, and
2. We know whether *any* request was successful, which determines if we'll call the handler with an error response.

If we could base `expected_result_count` on the number of clients and call `sync_handler` unconditionally, we could simplify the code:

```lua
function lsp.buf_request_all(bufnr, method, params, callback)
  bufnr = resolve_bufnr(bufnr)

  local clients = lsp.buf_get_clients(bufnr)
  local client_request_ids = {}

  local request_results = {}
  local result_count = 0
  local expected_result_count = vim.tbl_count(clients)
  local function _sync_handler(err, _, result, client_id)
    request_results[client_id] = { error = err, result = result }
    result_count = result_count + 1

    if result_count >= expected_result_count then
      callback(request_results)
    end
  end

  local make_unsupported_err = once(function()
    return lsp._unsupported_method(method)
  end)

  for _, client in ipairs(clients) do
    if client.supports_method(method) then
      local request_success, request_id = client.request(method, params, _sync_handler, bufnr)
      if request_success then
        client_request_ids[client.id] = request_id
      end
    else
      _sync_handler(make_unsupported_err(), method, nil, client.id)
    end
  end

  return function()
    for client_id, request_id in pairs(client_request_ids) do
      local client = active_clients[client_id]
      client.cancel_request(request_id)
    end
  end
end
```

**Comment by mjlbach** [comment]
Sorry, this is something we should address now.

I'm not comfortable with the change that requires scheduling the handler. One option would be to iterate over clients, get the suppported methods, and then modify for_each_buffer_client to take an optional client_id list (or have an alternative function) that we could then pass the client ids that support the method. Thoughts? @neovim/lsp

**Pull request closed without merging**